### PR TITLE
Blockmatrix fix

### LINKED ISSUE #1930: Problem with block matrix
I found several problems with the block matrix illustrated with the following example : 

```c++
#include "mfem.hpp"
#include <fstream>
#include <iostream>
using namespace std;
using namespace mfem;

int main(int argc, char *argv[])
{
    SparseMatrix *matA = new SparseMatrix(2, 2);
    matA->Set(0,0,11.);    matA->Set(0,1,12.);
    matA->Set(1,0,13.);    matA->Set(1,1,14.);

    SparseMatrix *matB = new SparseMatrix(2, 2);
    matB->Set(0,0,21.);    matB->Set(0,1,22.);
    matB->Set(1,0,23.);    matB->Set(1,1,24.);

    SparseMatrix *matC = new SparseMatrix(2, 2);
    matC->Set(0,0,31.);    matC->Set(0,1,32.);
    matC->Set(1,0,33.);    matC->Set(1,1,34.);

    SparseMatrix *matD = new SparseMatrix(2, 2);
    matD->Set(0,0,41.);    //matD->Set(0,1,42.);
    matD->Set(1,0,43.);    matD->Set(1,1,44.);

    Array<int> block_offsets(3);
    block_offsets[0] = 0;    block_offsets[1] = 2;    block_offsets[2] = 4;

    BlockMatrix *block_mat = new BlockMatrix(block_offsets,block_offsets);
    block_mat->SetBlock(0, 0, matA);    block_mat->SetBlock(0, 1, matB);
    block_mat->SetBlock(1, 0, matC);    block_mat->SetBlock(1, 1, matD);
    block_mat->Finalize();
    block_mat->Print();
}
```

I create a 2x2 block matrix formed by 2x2 elements sparse matrices.
First, the null block check in blockmatrix.cpp line 150 is using the element index instead of block index and throws an error when accessing the third row because there is only 2 blocks. The first problem is easy to fix :
`blockmatrix.cpp` :
```c++
const double& BlockMatrix::Elem (int i, int j) const
{
   int iloc, iblock;
   int jloc, jblock;

   findGlobalRow(i, iblock, iloc);
   findGlobalCol(j, jblock, jloc);

   if (IsZeroBlock(i, j)) // <<<<<<<    IsZeroBlock(iblock, jblock) should be used 
   {
      mfem_error("BlockMatrix::Elem");
   }

   return Aij(iblock, jblock)->Elem(iloc, jloc);
}
```

Then, the BlockMatrix::Print() method tries to access every element as if it was a dense matrix but if an element is missing (in my example, I removed matD(0,1)), the method throws an error because matD is really sparse and the element does not exist.
I don't know if this cause problems elsewhere and I'm not sure how to fix it.
Does it make sense to return the value 0 when trying to access an undefined sparse matrix element in the SparseMatrix::Elem(i,j) method ?
An other possibility could be to loop over the block, then loop over the block's row instead of a global loop.
Best regards
Gauthier

### PR DESCRIPTION
**Pull request by gauthier12**
Fix blockmatrix problems described in issue #1930

Resolves #1930.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1935](https://github.com/mfem/mfem/pull/1935) | @gauthier12 | @v-dobrev | @dylan-copeland + @homijan | 12/21/20 | 02/16/21 | 02/25/21 | |

### CODE REVIEW

**Comment by tzanio** [comment]
Hi @gauthier12 ,

When this is ready, can you please mark it with the `ready-for-review` label?

See https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#quick-summary

Tzanio

**Comment by gauthier12** [comment]
Hi @tzanio 
I didn't know for the `ready-for-review`. I marked it and I know for the next time.
Cheers 
Gauthier

**Comment by gauthier12** [comment]
I'm removing the label as there is discussion going on issue #1930

**Review by uvilla — COMMENTED** [review]
`SparseMatrix::Elem const` does not throw any error unless `i` or `j` are invalid (ie less than 0 or larger than the number of row/columns).
`BlockMatrix::Elem const` should do the same.

I think we should replace the assert at line 160 with `return 0.`

File: linalg/blockmatrix.cpp [hunk 1]
```diff
@@ -155,12 +155,11 @@ const double& BlockMatrix::Elem (int i, int j) const
    findGlobalRow(i, iblock, iloc);
    findGlobalCol(j, jblock, jloc);
 
-   if (IsZeroBlock(i, j))
+   if (IsZeroBlock(iblock, jblock))
    {
-      mfem_error("BlockMatrix::Elem");
+      return 0;
```
Comment (dylan-copeland) [inline linalg/blockmatrix.cpp:160]: This should return `zero`, with `static const double zero = 0.0;`

Comment (gauthier12) [reply linalg/blockmatrix.cpp]: Sorry for that. I pushed the correction

**Review by dylan-copeland — APPROVED** [review]
Thank you @gauthier12 for your contribution!

**Review by uvilla — APPROVED** [review]
Thank you @gauthier12

**Comment by v-dobrev** [comment]
Merged in `next` for testing...

**Comment by v-dobrev** [comment]
@homijan, I now realize you have not approved this PR. Did you want to take a look at it? We already have two approvals, so it should be good to go in general -- I just wanted to check with you.

**Pull request merged**